File: app/src/main/java/com/tutv/android/ui/series/SeriesPresenter.java
package com.tutv.android.ui.series;

import com.tutv.android.domain.Episode;
import com.tutv.android.domain.Season;
import com.tutv.android.domain.Series;
import com.tutv.android.repository.SeriesRepository;
import com.tutv.android.utils.schedulers.BaseSchedulerProvider;

import java.lang.ref.WeakReference;

import io.reactivex.Single;
import io.reactivex.disposables.CompositeDisposable;

public class SeriesPresenter {

    private final WeakReference<SeriesView> seriesView;

    private final SeriesRepository seriesRepository;
    private final BaseSchedulerProvider schedulerProvider;
    private final int seriesId;
    //Todo: el tratamiento del objeto series deberia ser a traves de un Optional, pero no estamos en ese api level...
    private Series series;


    private final CompositeDisposable disposables;

    public SeriesPresenter(SeriesView seriesView, int seriesId, SeriesRepository seriesRepository, BaseSchedulerProvider schedulerProvider) {
        this.seriesView = new WeakReference<>(seriesView);
        this.seriesId = seriesId;
        this.seriesRepository = seriesRepository;
        this.schedulerProvider = schedulerProvider;

        this.series = null;

        this.disposables = new CompositeDisposable();
    }

    public void onViewAttached() {
        disposables.add(seriesRepository.getSeriesById(seriesId)
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe(this::onSeriesLoad, this::onSeriesLoadError));
    }

    public void onViewDetached() {
        disposables.dispose();
    }

    private void onSeriesLoad(Series series) {
        this.series = series;
        SeriesView actualView = seriesView.get();
        if(actualView != null) {
            actualView.showSeriesName(series.getName());
            actualView.showSeriesDescription(series.getSeriesDescription());
            actualView.showSeriesFollowed(series.getLoggedInUserFollows() == null ? false : series.getLoggedInUserFollows());
            actualView.showFollowerCount(series.getFollowers());
            actualView.bindSeasons(series.getSeasons());
            actualView.showSeriesBanner(series.getBannerUrl());
        }
    }

    private void onSeriesLoadError(final Throwable e) {
        SeriesView view = seriesView.get();
        if(view != null) {
            view.showError("Error al cargar la serie, asegurese de tener conexion");
        }
    }

    public void onEpisodeClicked(Season s, Episode e) {
        disposables.add(seriesRepository.setEpisodeViewed(series, s, e)
                .observeOn(schedulerProvider.computation())
                .flatMap(series -> {
                    for(Season season : series.getSeasons()) {
                        if(season.getNumber() == s.getNumber()) {
                            return Single.just(s);
                        }
                    }
                    return Single.error(new Throwable());
                })
                .observeOn(schedulerProvider.ui())
                .subscribe(this::onEpisodeViewed, this::onEpisodeViewedError));
    }

    private void onEpisodeViewedError(Throwable throwable) {
        System.err.println(throwable);
        SeriesView view = seriesView.get();
        if (view != null) {
            view.showError("Error al ver el episodio, asegurese de tener conexion y estar logueado");
        }
    }

    private void onEpisodeViewed(Season season) {
        SeriesView view = seriesView.get();
        if(view != null) {
            view.bindSeason(season);
        }
    }

    public void onSeriesFollowClicked() {
        if(series == null) return;
        if(series.getLoggedInUserFollows() == null || !series.getLoggedInUserFollows()) {
            disposables.add(seriesRepository.setFollowSeries(series)
                    .observeOn(schedulerProvider.ui())
                    .subscribe(this::onSeriesFollowed, this::onSeriesFollowedError));
        } else {
            disposables.add(seriesRepository.unfollowSeries(series)
                    .observeOn(schedulerProvider.ui())
                    .subscribe(this::onSeriesUnfollowed, this::onSeriesUnfollowedError));
        }
    }

    private void onSeriesUnfollowedError(Throwable throwable) {
        SeriesView view = seriesView.get();
        if(view != null) {
            view.showError("Error al dejar de seguir la serie, asegurese de tener conexion y estar logueado");
        }
    }

    private void onSeriesUnfollowed(Series series) {
        if(series == null) return;
        SeriesView view = seriesView.get();
        if(view != null) {
            view.showSeriesFollowed(false);
            view.showFollowerCount(series.getFollowers());
        }
    }

    private void onSeriesFollowedError(Throwable throwable) {
        SeriesView view = seriesView.get();
        if(view != null) {
            view.showError("Error al seguir la serie, asegurese de tener conexion y estar logueado");
        }
    }

    private void onSeriesFollowed(Series series) {
        if(series == null) return;
        SeriesView view = seriesView.get();
        if(view != null) {
            view.showFollowerCount(series.getFollowers());
            view.showSeriesFollowed(series.getLoggedInUserFollows());
        }
    }
}

File: app/src/main/java/com/tutv/android/ui/series/SeriesPresenter.kt
package com.tutv.android.ui.series

import com.tutv.android.domain.*
import com.tutv.android.repository.SeriesRepository
import com.tutv.android.utils.schedulers.BaseSchedulerProvider
import io.reactivex.Single
import io.reactivex.disposables.CompositeDisposable
import java.lang.ref.WeakReference

class SeriesPresenter(
        seriesView: SeriesView?,
        var seriesId: Int,
        var seriesRepository: SeriesRepository,
        var schedulerProvider: BaseSchedulerProvider) {

    private val seriesView: WeakReference<SeriesView?> = WeakReference(seriesView)
    private var series: Series? = null
    private var reviews: MutableList<Review>? = null
    private val disposables: CompositeDisposable = CompositeDisposable()

    fun onViewAttached() {
        disposables.add(seriesRepository.getSeriesById(seriesId)
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe({ series: Series? -> onSeriesLoad(series) }) { e: Throwable? -> onSeriesLoadError(e) })

        disposables.add(seriesRepository.getReviews(seriesId)
                .subscribeOn(schedulerProvider.io())
                .observeOn(schedulerProvider.ui())
                .subscribe({ reviews: List<Review> -> onReviewsLoad(reviews) }) { e: Throwable? -> onReviewsLoadError(e) })
    }

    fun onViewDetached() {
        disposables.dispose()
    }

    private fun onSeriesLoad(series: Series?) {
        this.series = series
        val actualView = seriesView.get()
        if (actualView != null) {
            actualView.showSeriesName(series?.name)
            actualView.showSeriesDescription(series?.seriesDescription)
            actualView.showSeriesFollowed(series?.loggedInUserFollows ?: false)
            actualView.showFollowerCount(series?.followers)
            actualView.bindSeasons(series?.seasons)
            actualView.showSeriesBanner(series?.bannerUrl)
        }
    }

    private fun onReviewsLoad(reviews: List<Review>?) {
        this.reviews = reviews?.toMutableList()
        seriesView.get()?.bindReviews(this.reviews)
    }

    private fun onSeriesLoadError(e: Throwable?) {
        val view = seriesView.get()
        view?.showError("Error al cargar la serie, asegurese de tener conexion")
    }

    private fun onReviewsLoadError(e: Throwable?) {
        val view = seriesView.get()
        view?.showError("Error al cargar los comentarios, asegurese de tener conexion")
    }

    fun onEpisodeClicked(s: Season, e: Episode) {
        if (series == null) return

        disposables.add(seriesRepository.setEpisodeViewed(series!!, s, e)
                .observeOn(schedulerProvider.computation())
                .flatMap { series: Series? ->
                    for (season in series?.seasons!!) {
                        if (season.number == s.number) {
                            return@flatMap Single.just(s)
                        }
                    }
                    Single.error(Throwable())
                }
                .observeOn(schedulerProvider.ui())
                .subscribe({ season: Season? -> onEpisodeViewed(season) }) { throwable: Throwable? -> onEpisodeViewedError(throwable) })
    }

    private fun onEpisodeViewedError(throwable: Throwable?) {
        seriesView.get()?.showError("Error al ver el episodio, asegurese de tener conexion y estar logueado")
    }

    private fun onEpisodeViewed(season: Season?) {
        seriesView.get()?.bindSeason(season)
    }

    fun onSeriesFollowClicked() {
        if (series == null) return

        if (series!!.loggedInUserFollows == null || !series?.loggedInUserFollows!!) {
            disposables.add(seriesRepository.setFollowSeries(series!!)
                    .observeOn(schedulerProvider.ui())
                    .subscribe({ series: Series? -> onSeriesFollowed(series) }) { throwable: Throwable? -> onSeriesFollowedError(throwable) })
        } else {
            disposables.add(seriesRepository.unfollowSeries(series!!)
                    .observeOn(schedulerProvider.ui())
                    .subscribe({ series: Series? -> onSeriesUnfollowed(series) }) { throwable: Throwable? -> onSeriesUnfollowedError(throwable) })
        }
    }

    private fun onSeriesUnfollowedError(throwable: Throwable?) {
        seriesView.get()?.showError("Error al dejar de seguir la serie, asegurese de tener conexion y estar logueado")
    }

    private fun onSeriesUnfollowed(series: Series?) {
        if (series == null) return
        val view = seriesView.get()
        if (view != null) {
            view.showSeriesFollowed(false)
            view.showFollowerCount(series.followers)
        }
    }

    private fun onSeriesFollowedError(throwable: Throwable?) {
        seriesView.get()?.showError("Error al seguir la serie, asegurese de tener conexion y estar logueado")
    }

    private fun onSeriesFollowed(series: Series?) {
        if (series == null) return
        val view = seriesView.get()
        if (view != null) {
            view.showFollowerCount(series.followers)
            view.showSeriesFollowed(series.loggedInUserFollows ?: false)
        }
    }

    fun onReviewLikeClicked(r: Review) {
        if (series == null) return

        disposables.add(seriesRepository.setReviewLiked(series!!, r, !(r.loggedInUserLikes ?: false))
                .observeOn(schedulerProvider.ui())
                .subscribe({ review: Review? -> onReviewLiked(review) }) { throwable: Throwable? -> onReviewLikedError(throwable) })
    }

    private fun onReviewLiked(r: Review?) {
        seriesView.get()?.bindReviews(this.reviews)
    }

    private fun onReviewLikedError(throwable: Throwable?) {
        seriesView.get()?.showError("Error al dar me gusta al comentario, asegurese de tener conexion y estar logueado")
    }

    fun onReviewSubmitClicked(body: String?) {
        if (series == null || body.isNullOrEmpty()) return

        disposables.add(seriesRepository.postReview(series!!, body)
                .observeOn(schedulerProvider.ui())
                .subscribe({ review: Review? -> onReviewSubmitted(review) }) { throwable: Throwable? -> onReviewSubmittedError(throwable) })
    }

    private fun onReviewSubmitted(r: Review?) {
        if (r != null) {
            this.reviews?.add(0, r)
            seriesView.get()?.bindReviews(this.reviews)
        }
    }

    private fun onReviewSubmittedError(throwable: Throwable?) {
        seriesView.get()?.showError("Error al publicar el comentario, asegurese de tener conexion y estar logueado")
    }

}